python_modules/dagster/dagster/_core/definitions/asset_checks.py

```diff
@@ -127,3 +143,66 @@ def get_attributes_dict(self) -> Dict[str, Any]:
             specs=self._specs,
             input_output_props=self._input_output_props,
         )
+
+
+@experimental
+def build_blocking_asset_check(
+    asset_def: "AssetsDefinition",
+    checks: Sequence[AssetChecksDefinition],
+) -> "AssetsDefinition":
+    from dagster import AssetIn, In, OpExecutionContext, Output, graph_asset, op
+    from dagster._core.storage.asset_check_execution_record import AssetCheckExecutionRecordStatus
+
+    check_specs = []
+    for c in checks:
+        check_specs.extend(c.specs)
+
+    check_output_names = [c.get_python_identifier() for c in check_specs]
+
+    @op(ins={"materialization": In(Any), "check_evaluations": In(Nothing)})
+    def fan_in_checks_and_materialization(context: OpExecutionContext, materialization):
+        yield Output(materialization)
+
+        for check_spec in check_specs:
+            executions = context.instance.event_log_storage.get_asset_check_executions(
+                asset_key=asset_def.key, check_name=check_spec.name, limit=1
+            )
+            check.invariant(
+                len(executions) == 1, "Expected asset check {check_spec.name} to execute"
+            )
+            execution = executions[0]
+            check.invariant(
+                execution.run_id == context.run_id,
+                "Expected asset check {check_spec.name} to execute in the current run",
+            )
+            if execution.status != AssetCheckExecutionRecordStatus.SUCCEEDED:
+                raise DagsterAssetCheckFailedError()
+
+    @graph_asset(
+        name=asset_def.key.path[-1],
+        key_prefix=asset_def.key.path[:-1] if len(asset_def.key.path) > 1 else None,
+        group_name=asset_def.group_names_by_key.get(asset_def.key),
+        partitions_def=asset_def.partitions_def,
+        check_specs=check_specs,
+        description=asset_def.descriptions_by_key.get(asset_def.key),
+        ins={name: AssetIn(key) for name, key in asset_def.keys_by_input_name.items()},
+        resource_defs=asset_def.resource_defs,
+        metadata=asset_def.metadata_by_key.get(asset_def.key),
+        freshness_policy=asset_def.freshness_policies_by_key.get(asset_def.key),
```

Comment:
I don't really like this- easy to miss a field, annoying to test them all. Any recs?

Reply:
In terms of easy to miss a field and for future proofing, the only way I can really think of is to refactor `graph_asset` to immediately all a function `graph_asset_no_defaults` and then write test cases against `graph_asset_no_defaults`. That way anyone who adds something in the future will break all the tests, which is the desired behavior, as the right fix will undoubtedly be to change some code here.

Reply:
On the "test them all" front, need more context on what you need/want to test

Reply:
tests that assert these parameters are threaded through. 👍 to graph_asset_no_defaults